tools/discovery/regexp.go

```diff
@@ -0,0 +1,67 @@
+package discovery
+
+import (
+	"regexp"
+)
+
+var (
+	// TODO: This doesnot conform to https://tools.ietf.org/html/rfc3986 yet.
+	host     = match(`[a-zA-Z0-9][a-zA-Z0-9\:\_\-\.]{1,128}`)
+	path     = match(`[a-z0-9-_]+(?:/[a-z0-9-_.]+){0,7}`)
+	fragment = match(`[a-z0-9-_.]{1,128}`)
+
+	RegHostBasedImage = anchored(capture(host), literal(`/`), capture(path), optional(literal(`#`), capture(fragment)))
```

Comment:
I [still][1] think you want this regexp to be private, and that you [don't need the helper functions][2] to build it.

[1]: https://github.com/xiekeyang/oci-discovery/commit/ce7083d030142ac71884cb4b1a95bdadab4eace9#commitcomment-24312383
[2]: https://github.com/xiekeyang/oci-discovery/commit/ce7083d030142ac71884cb4b1a95bdadab4eace9#commitcomment-24312309

Reply:
> I still think you want this regexp to be private…

#17 has a fixup for this PR which does this.

Reply:
> I [still][1] think you want this regexp to be private, and that you [don't need the helper functions][2] to build it.

I feel helper functions is needed. Later we might expend more `regexps`, some `regexp` might be group of another. helper make expend easy, and make easy to modify some `regexp` (maybe just need modify one segment in a long `regexp`). It is easy to read. So I suggest we keep it. At least currently I see it make me easy to understand, and reduce the risk of syntax error.

Reply:
> #17 has a fixup for this PR which does this.

I guess [1] is correct, but I think it is a little hard to read, and hard to modify. Please consider to keep helper functions. 

[1]: https://github.com/wking/oci-discovery/blame/e3e42311f81c08bacf55d26938951ae8c3071dd1/tools/hostbasedimagenames/main.go#L8-L18

Reply:
But helper function have no `group` name, should be added.

Reply:
> I guess [1][] is correct, but I think it is a little hard to read, and hard to modify. Please consider to keep helper functions.

I tried to stick to the RFC's ABNF, with a private variable corresponding to each ABNF rule.  Let me know if you want more details on any particular rule.

And it's your project.  If you find the helper-function approach easier to read, feel free to keep it.

[1]: https://github.com/wking/oci-discovery/blame/e3e42311f81c08bacf55d26938951ae8c3071dd1/tools/hostbasedimagenames/main.go#L8-L18